File: problems/House Robber/rob.cs
public class Solution {
    public int Rob(int[] nums) {
        if (null == nums || 0 == nums.Length) {
            return 0;
        }

        int prev = nums[0];
        int prevPrev = 0;

        for (int i = 1; nums.Length > i; ++i) {
            int curr = Math.Max(nums[i] + prevPrev, prev);
            prevPrev = prev;
            prev = curr;
        }

        return prev;
    }
}

File: problems/House Robber/rob.java
class Solution {
    public int rob(int[] nums) {
        Integer[] dp = new Integer[nums.length];

        return rodTopDown(nums, dp, 0);
    }

    private int rodTopDown(int[] nums, Integer[] dp, int idx) {
        if (nums.length <= idx) {
            return 0;
        }

        if (null == dp[idx]) {
            dp[idx] = Math.max(nums[idx] + rodTopDown(nums, dp, 2 + idx), rodTopDown(nums, dp, 1 + idx));
        }

        return dp[idx];
    }
}
